spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -281,35 +297,84 @@ public void onPartitionsAssigned(Map<TopicPartition, Long> assignments, Consumer
 		public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment, Consumer<?, ?> consumer) {
 			Message<?> message = null;
 			try {
-				message = enhanceHeaders(toMessagingMessage(record, acknowledgment, consumer), record);
-				setAttributesIfNecessary(record, message);
+				message = toMessagingMessage(record, acknowledgment, consumer);
 			}
 			catch (RuntimeException e) {
+				if (KafkaInboundGateway.this.retryTemplate == null) {
+					message = enhanceHeaders(message, record);
+				}
 				MessageChannel errorChannel = getErrorChannel();
 				if (errorChannel != null) {
 					KafkaInboundGateway.this.messagingTemplate.send(errorChannel, buildErrorMessage(null,
 							new ConversionException("Failed to convert to message", record, e)));
 				}
 			}
 			if (message != null) {
-				try {
-					Message<?> reply = sendAndReceiveMessage(message);
-					if (reply != null) {
-						reply = enhanceReply(message, reply);
-						KafkaInboundGateway.this.kafkaTemplate.send(reply);
-					}
-				}
-				finally {
-					if (KafkaInboundGateway.this.retryTemplate == null) {
-						ATTRIBUTES_HOLDER.remove();
-					}
-				}
+				sendAndReceive(record, message, acknowledgment, consumer);
 			}
 			else {
 				KafkaInboundGateway.this.logger.debug(() -> "Converter returned a null message for: " + record);
 			}
 		}
 
+		private void sendAndReceive(ConsumerRecord<K, V> record, Message<?> message, Acknowledgment acknowledgment,
+				Consumer<?, ?> consumer) {
+
+			RetryTemplate template = KafkaInboundGateway.this.retryTemplate;
+			if (template != null) {
+				doWithRetry(template, KafkaInboundGateway.this.recoveryCallback, record, acknowledgment, consumer,
+						() -> {
+							Message<?> toSend = enhanceHeaders(message, record, acknowledgment, consumer);
+							if (toSend != null) {
+								doSendAndReceive(message);
+							}
+						});
+			}
+			else {
+				Message<?> toSend = enhanceHeaders(message, record, acknowledgment, consumer);
+				if (toSend != null) {
+					doSendAndReceive(toSend);
+				}
+			}
+		}
+
+		@Nullable
+		private Message<?> enhanceHeaders(Message<?> message, ConsumerRecord<K, V> record,
+				Acknowledgment acknowledgment, Consumer<?, ?> consumer) {
+
+			try {
+				Message<?> message2 = enhanceHeaders(message, record);
+				setAttributesIfNecessary(record, message2);
+				return message2;
+			}
+			catch (RuntimeException ex) {
+				MessageChannel errorChannel = getErrorChannel();
+				if (errorChannel != null) {
+					KafkaInboundGateway.this.messagingTemplate.send(errorChannel, buildErrorMessage(null,
+							new ConversionException("Failed to convert to message", record, ex)));
+					return null;
```

Comment:
However this may cover us.
Please, confirm that this is an intention for the missed `else` I mention before.

Reply:
See comment above.